### PR TITLE
Add practical support for transmitter handling and position detection.

### PR DESCRIPTION
**Pull request by djpetti**
This includes updates to the Kalman Filter code for improved robustness, reliability, and stability, as well as all the code for observing and calculating the position of transmitters based on radio data. Also included are some standard unit tests as well as a full simulation-based test suite for assessing the higher-level performance of the code without needing to run it on a drone.

### CODE REVIEW

File: belief_manager.py [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+
+    # How many cycles we've run.
+    self._cycles = 0
+
+  def _fetch_autopilot_data(self):
+    """ Fetches measurements from the onboard Pixhawk autopilot, and updates the
+    member variables for vehicle position and velocity observations accordingly.
+    """
+    # Read the latest data from the autopilot.
+    sensor_data = self.__autopilot.read_front()
```
Comment (wkronmiller) [inline belief_manager.py]: What if there is no new data since last read?

File: belief_manager.py [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+  # How far away (m) we can be from a transmitter signal and still register it,
+  # in average conditions.
+  RADIO_RANGE = 40.0
+
+  def __init__(self):
+    self._initialize_member_variables()
+
+    # Initialize connections to the Pixhawk reader and wireless
+    # communications handler, as well as the transmitter locator.
+    factory = SocketIPCFactory()
```
Comment (wkronmiller) [inline belief_manager.py]: Will have to re-factor this later to use Multiprocessing. I suggest taking queues you need as arguments to constructor.

File: belief_manager.py [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+      signal = self.__radio.read_next(block=False)
+      if not signal:
+        # No more data.
+        break
+
+      # Normalize the LOB, so that the angle will still be correct even if the
+      # drone starts flying in a different direction.
+      nomalized_lob = self._filter.normalize_lobs(signal.lob)
+
+      transmitters.append((normalized_lob, signal.strength))
```
Comment (wkronmiller) [inline belief_manager.py]: What if there are multiple transmissions from same source in queue?

Comment (wkronmiller) [inline belief_manager.py]: How does this account for craft position at time of reading?

File: belief_manager.py [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+      and the values are the tuple of the associated LOB and signal strength. It
+      also returns a list of readings that are asumed to be from new
+      transmitters. """
+    margins_of_error = \
+        self._filter.lob_confidence_intervals(self.ERROR_REGION_Z_SCORE)
+
+    # Center the interval around our predicted next LOBs.
+    predicted_state = self._filter.estimate_next_state()
+    intervals = []
+    for i in range(4, len(predicted_state)):
```
Comment (wkronmiller) [inline belief_manager.py]: What is 4?

File: belief_manager.py [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+      lob = predicted_state[i]
+      margin_of_error = margins_of_error[i - 4]
+      intervals.append((lob - margin_of_error, lob + margin_of_error))
+
+    logger.debug("LOB confidence intervals: %s" % (intervals))
+
+    # Now, go and check whether our new LOBs fit within them.
+    associations = {}
+    new = []
```
Comment (wkronmiller) [inline belief_manager.py:162]: new_?

File: belief_manager.py [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+    for reading in readings:
+      associated = False
+      weak = False
+      best_center_distance = sys.maxint
+      best_transmitter = 0
+
+      lob = reading[0]
```
Comment (wkronmiller) [inline belief_manager.py]: Magic! Perhaps arrays aren't the best data structure to use for this?

File: belief_manager.py [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+      for i in range(0, len(intervals)):
+        transmitter = i + 4
+        lower_bound = intervals[i][0]
+        upper_bound = intervals[i][1]
+        if (lob >= lower_bound and lob <= upper_bound):
+          # It's in range.
+          old_association = associations.get(transmitter)
+          if old_association:
+            # If this one is a stronger signal, replace it.
+            if (strength <= old_association[0]):
```
Comment (wkronmiller) [inline belief_manager.py]: What is 0?

File: belief_manager.py [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+          self._cycle_data[transmitter] = self._cycles
+          associated = True
+          best_center_distance = center_distance
+          best_transmitter = transmitter
+
+      if (not associated and not weak):
+        # It fit inside none of our previous regions.
+        logger.info("Asuming %s is new transmitter." % (str(reading)))
+        new.append(reading)
+        new_index = 4 + self._filter.number_of_transmitters() + len(new)
```
Comment (wkronmiller) [inline belief_manager.py]: What is this 4?

File: belief_manager.py [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+    logger.debug("Associated bearings with transmitters: %s" % (associations))
+    return (associations, new)
+
+  def _distance_from_strength(self, readings):
+    """ Takes a set of transmitter readings along with their strengths and uses
+    their strengths to guess how far away they are. This is not a very accurate
+    method of guaging transmitter positions, but it is used initially, because
+    it works with only one reading.
+    Args:
+      readings: A set of associated LOB readings, or a list of LOB, strength
```
Comment (wkronmiller) [inline belief_manager.py:223]: Why both?

File: belief_manager.py [hunk 10]
```diff
@@ -0,0 +1,578 @@
...
+    transmitters: The indexes of the transmitters in the state to use.
+    Returns: A dict of tuples, with each tuple containing the estimated x and y
+      position of the transmitter. The keys are the indices of the corresponding
+      transmitters in the state. """
+    logger.debug("Past lobs: Calculating distance.")
+    positions = {}
+    for transmitter in transmitters:
+      # Check whether we have enough data to use the filtered state.
+      current_lob = self._filter.state()[transmitter]
+      current_position = self._filter.position()
```
Comment (wkronmiller) [inline belief_manager.py]: Does this update between iterations?

File: belief_manager.py [hunk 11]
```diff
@@ -0,0 +1,578 @@
...
+    """
+    # To find out where they cross, we first have to translate the lobs into
+    # actual lines. Start with the current one.
+    current_slope = np.tan(new_lob)
+    current_intercept = position[self._Y] - current_slope * \
+                        position[self._X]
+
+    # Translate the old one.
+    past_data = self._past_lobs[transmitter]
+    past_lob = past_data[0]
```
Comment (wkronmiller) [inline belief_manager.py]: Magic

File: belief_manager.py [hunk 12]
```diff
@@ -0,0 +1,578 @@
...
+    return distances
+
+  def _transmitter_error_regions(self, stddevs, existing, new):
+    """ Calculates error regions for the transmitters that we can currently see,
+    and saves them.
+    Args:
+      stddevs: The z-score to use when calculating the error regions.
+      existing: Associated readings for existing transmitters.
+      new: Readings for new transmitters. """
+    num_points = 40 + 10 * self._filter.number_of_transmitters()
```
Comment (wkronmiller) [inline belief_manager.py]: Magic

File: belief_manager.py [hunk 13]
```diff
@@ -0,0 +1,578 @@
...
+    # For each transmitter, start building up transformed sets of points that
+    # describe the error regions around that particular transmitter.
+    new_index = 0
+    for i in range(4, len(self._filter.state())):
+      if (i - 4 >= len(self._past_lobs)):
+        # It's new enough that we're going to have to rely on the strength
+        # calculations instead.
+        logger.debug("Falling back on strength calculator for %d." % (i))
+
+        if i >= len(self._filter.state()) - len(new):
```
Comment (wkronmiller) [inline belief_manager.py:387]: Could you double-check this math, then further explain what it does?

File: belief_manager.py [hunk 14]
```diff
@@ -0,0 +1,578 @@
...
+          if not strength:
+            # We don't have any data on this transmitter, so our hands are
+            # kind of tied.
+            logger.warning("Skipping transmitter %d with no data." % (i))
+            continue
+          strength = strength[1]
+
+      transformed_points = []
+      for point in error_points:
+        hypothetical_lob = point[i]
```
Comment (wkronmiller) [inline belief_manager.py]: Again, some other data structure than an array might be more appropriate.

File: belief_manager.py (hunk 1 above)
Comment (djpetti) [inline belief_manager.py]: Done.

File: belief_manager.py (hunk 2 above)
Comment (djpetti) [inline belief_manager.py]: Done.

File: belief_manager.py (hunk 3 above)
Comment (djpetti) [inline belief_manager.py]: 1. The _associate_lobs method is designed to handle this case.
2. For now, we're assuming that velocity remains constant, i.e. the drone is always pointing in the same direction. (Handling changing velocities will be coming at a future time...) I'll put a TODO in there reminding me to change it when the time comes, though.

File: belief_manager.py (hunk 5 above)
Comment (djpetti) [inline belief_manager.py:162]: Uhh... I don't see why the name needs to change. It's not like "new" is a protected name in Python.

File: belief_manager.py (hunk 4 above)
Comment (djpetti) [inline belief_manager.py]: Done.

File: belief_manager.py (hunk 6 above)
Comment (djpetti) [inline belief_manager.py]: As per your above comment, I got rid of the magic numbers here too, which should hopefully make things a little clearer. As far as the question of data structures, I'm not completely sure what you're referring to, seeing as you've selected an entire block of code.

File: belief_manager.py (hunk 8 above)
Comment (djpetti) [inline belief_manager.py]: Done.

File: belief_manager.py (hunk 7 above)
Comment (djpetti) [inline belief_manager.py]: Done.

File: belief_manager.py (hunk 9 above)
Comment (djpetti) [inline belief_manager.py:223]: Done.

File: belief_manager.py (hunk 10 above)
Comment (djpetti) [inline belief_manager.py]: Once again, it's not very clear what exactly you're referring to. I'm going to assume it's the bottom line, in which case, it does change every iteration. All the position() method does is it goes and pulls out the position from the filtered state in the Kalman filter. The comment above it, however, is outdated and confusing, and I have taken the liberty of fixing it.

File: belief_manager.py (hunk 11 above)
Comment (djpetti) [inline belief_manager.py]: There is a comment somewhere that explains exactly how _past_lobs is organized. I have, however, extended the comment in this immediate area in order to make it completely clear.

File: belief_manager.py (hunk 12 above)
Comment (djpetti) [inline belief_manager.py]: Yeah, can't argue with that... I think it's mainly lacking a comment, so I've added one.

File: belief_manager.py (hunk 13 above)
Comment (djpetti) [inline belief_manager.py:387]: I'm pretty sure this is correct, especially considering that most of the actual work is farmed out to other methods that work correctly in other places. Also, it is unit-tested...

File: belief_manager.py (hunk 14 above)
Comment (djpetti) [inline belief_manager.py]: Each point on the error ellipsoid has the same layout as the state. The state is represented as an array, therefore, I don't see why an array is a poor choice for representing points on the error ellipsoid.

File: belief_manager.py (hunk 5 above)
Comment (wkronmiller) [inline belief_manager.py:162]: My bad. Unclear message. I meant to indicate you should rename the variable to new_[something that gives context].

File: belief_manager.py [hunk 15]
```diff
@@ -0,0 +1,621 @@
...
+                         (transmitter[self._Y] - position[self._Y] ** 2))
+      if distance <= self.RADIO_RANGE:
+        visible_transmitters.append(transmitter)
+
+    # Check if we are seeing fewer of the visible transmitters than we would
+    # expect.
+    for num_transmitters in self._observed_transmitters:
+      if num_transmitters >= len(visible_transmitters):
+        break
+    else:
```
Comment (wkronmiller) [inline belief_manager.py]: Not a bug, just blindness: What "if" is this else associated with?

File: belief_manager.py (hunk 5 above)
Comment (djpetti) [inline belief_manager.py:162]: Ahhh. Okay. That makes more sense.

**Comment by wkronmiller** [comment]
Awesome

**Pull request merged**